controllers/baremetalset_controller.go

```diff
@@ -274,32 +274,67 @@ func (r *BaremetalSetReconciler) ensureBaremetalHosts(instance *ospdirectorv1bet
 			r.Log.Info(fmt.Sprintf("Existing BaremetalHost: %s", baremetalHost.ObjectMeta.Name))
-			existingBaremetalHosts = append(existingBaremetalHosts, baremetalHost.ObjectMeta.Name)
+			// Storing the existing BaremetalHosts in a map like this just makes selected replica scale-down easier below
+			existingBaremetalHosts[baremetalHost.ObjectMeta.Name] = baremetalHost.ObjectMeta.Name
+
+			// Prep for possible replica scale-down below
+			if val, ok := baremetalHost.Annotations[baremetalset.BaremetalHostRemovalAnnotation]; ok && (val == "yes" || val == "true") {
+				removalAnnotatedBaremetalHosts = append(removalAnnotatedBaremetalHosts, baremetalHost.ObjectMeta.Name)
+			}
 		}
 	}
 
-	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.
-	// TODO: Add something to CR to allow user to specify which BaremetalHosts are removed
-	for i := len(existingBaremetalHosts); i > instance.Spec.Replicas; i-- {
-		for _, oldBmhName := range existingBaremetalHosts {
-			// Just delete the first one we find until the TODO from above is implemented
-			err := r.baremetalHostDeprovision(instance, oldBmhName)
+	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.  First we
+	// choose BaremetalHosts with the "osp-director.openstack.org/baremetalset-delete-baremetalhost=yes"
+	// annotation, then, if there are still BaremetalHosts left to deprovision, we choose randomly from the
+	// existingBaremetalHosts map (as maps are unordered in Golang, we don't know which one will be chosen)
+
+	// How many new BaremetalHost de-allocations do we need (if any)?
+	oldBmhsToRemoveCount := len(existingBaremetalHosts) - instance.Spec.Replicas
+
+	for i := 0; i < oldBmhsToRemoveCount; i++ {
+		// First choose BaremetalHosts to remove from the prepared list of BaremetalHosts
+		// that have the "osp-director.openstack.org/baremetalset-delete-baremetalhost=yes" annotation
+
+		if len(removalAnnotatedBaremetalHosts) > 0 {
+			err := r.baremetalHostDeprovision(instance, removalAnnotatedBaremetalHosts[0])
 
 			if err != nil {
 				return err
 			}
 
-			if len(existingBaremetalHosts) > 1 {
-				existingBaremetalHosts = existingBaremetalHosts[1:]
+			// Remove the removal-annotated BaremetalHost from the existingBaremetalHosts map
+			delete(existingBaremetalHosts, removalAnnotatedBaremetalHosts[0])
+
+			// Remove the removal-annotated BaremetalHost from the removalAnnotatedBaremetalHosts list
+			if len(removalAnnotatedBaremetalHosts) > 1 {
+				removalAnnotatedBaremetalHosts = removalAnnotatedBaremetalHosts[1:]
 			} else {
-				existingBaremetalHosts = []string{}
+				removalAnnotatedBaremetalHosts = []string{}
+			}
+
+			// We removed a removal-annotated BaremetalHost, so continue
+			continue
+		}
+
+		// We did NOT find a removal-annotated BaremetalHost to remove above, so we pick the first one
+		// in the existingBaremetalHosts list
```

Comment:
We probably should not do any action if the host to remove was not annotated for removal to make sure we do not remove a node which is not expected to be deleted. The workflow today in tripleo also requires specify a specific node to delete.